New retriever flow

rvagg opened the pull request:
Goal:

1. Implement a new retrieval flow such that we get parallel queries and then start on retrievals as soon as we have queries available so we're not blocking for _all_ of them. In addition, if we have multiple queries waiting in the queue at the same time, we prioritise and fetch from the best one next.
2. Make this whole thing more decoupled and therefore testable.

How:

1. I implemented a new `CidRetrieval` type to work on this new flow separately to the existing since the flow is very different. In the process I abstracted away a lot of the metrics reporting and active retrievals management stuff and a few other pieces that require dependencies I didn't want to push down that low. There are now callbacks and other typed functions that need to be provided to set up a retrieval.
2. I made a `PriorityWaitQueue` that can handle the blocking and sorting operation amongst goroutines without the need for an additional goroutine to do the sorting. So when queries are finished, they just need to `Wait()` their turn.

kylehuntsman commented:
I made some potential edits along with some of my thoughts in #13.

rvagg commented:
Ready for some review now.

@kylehuntsman I found that it wasn't hard to add a bit of extra functionality to cancel ongoing queries when we got a successful retrieval; so that's in here now and as a bonus I get to watch for cleaned up goroutines. Although, my cancellation is currently only exercised in the tests via the mock client but I believe the cancellation happens properly through to libp2p on that path because we use the same mechanism for timeouts already.

rvagg commented:
Updated to allow multiple retrievals per CidRetrieval instance, so now you can set up one and call `RetrieveCid()` on it multiple times, across different goroutines, and it'll work. This may or may not be useful in our current workflows but it does clean things up a little internally and isolates state to a single retrieval execution. @kylehuntsman this is kind of along the lines of one of the things you were trying to achieve in #13.

kylehuntsman commented on pkg/retriever/cidretrieval.go:
```diff
@@ -0,0 +1,502 @@
+			// have we got all responses but no success?
+			finishedCount++
+			if finishedCount >= len(candidates) {
+				break collectcomplete
```
Is this so that it breaks out of the loop and not the select?

rvagg replied:
yeah .. so the original version of this was a `for` across the channel, which was really clean, but then I realised I needed to be able to handle context cancellation, and I don't believe there's a good way I can do that, aside perhaps from closing the channel but that has other consequences _and_ that channel needs to be closed from some other goroutine! .. so this is it, and it's ugly, and it's the second block of code like this in this file now :shrug:.

hannahhoward commented on pkg/retriever/cidretrieval.go:
```diff
@@ -0,0 +1,502 @@
+
+// NewCidRetrieval creates a new CidRetrieval. A CidRetrieval instance may be
+// used for multiple retrievals via the RetrieveCid call.
+func NewCidRetrieval(
```
Public function returning private type is a bit of an anti-pattern. Any reason it's a private struct when all it's members are public?

rvagg replied:
ok, so should I switch it around, so the type is public but the fields are private? I wanted it all to be private but the "public" methods on it were just a way of making them stand out in the code, a weak attempt at making the code less ugly; but I don't think it achieved much.

kylehuntsman commented on pkg/retriever/cidretrieval.go:
```diff
@@ -0,0 +1,502 @@
+			// signals to goroutines to bail, this has to be done here, rather than on
+			// the receiving parent end, because immediately after this call we instruct
+			// the prioritywaitqueue that we're done and another may start
+			close(state.FinishChan)
```
Mmmm, at first glance this goes against what I've read about closing channels from N senders. Let's talk about this.

kylehuntsman replied:
Oh I see, you're relying on the priority queue named WaitQueue to ensure you only do one retrieval at a time.

rvagg replied:
yeah, I move this here because my tests showed a case where there's a small gap between the waitqueue being told we're done with one, and the result being received, allowing the next one to pick it up; so the `close()` can only happen while the rest are blocked

rvagg commented on pkg/retriever/cidretrieval.go:
```diff
@@ -0,0 +1,502 @@
+	WaitQueue           prioritywaitqueue.PriorityWaitQueue[*retrievalmarket.QueryResponse]
+	ResultChan          chan runResult
+	FinishChan          chan struct{}
+	ActiveQueryCancel   []*func()
```
it's for removing afterward, can't check a `fn == fn`, Go errors on that but pointer gives us an identity, see where this is used later in the file

rvagg commented on pkg/retriever/cidretrieval.go:
```diff
@@ -0,0 +1,502 @@
+	var cancelled bool
+	var cancelFunc func()
+	ctx, cancelFunc = context.WithCancel(ctx)
+	activeCancel := func() {
```
yes! in an ideal world .. but we also have the timeout context, and we need to be able to distinguish between "cancelled internally" and "timed out" so we return the right error so the "timed out" case bubbles back up the user and the cancelled case doesn't

So we could check the context, but we'd need to hold on to the two separate contexts .. which is fine, but is that any better than a bool? I just made a line judgement on which new variable to hold on to.

rvagg replied:
(also complexifying it is that the second context, for timeout, is optional, so hidden in an `if` just below that, so holding on to two separate contexts becomes a challenge)

kylehuntsman commented on pkg/retriever/cidretrieval.go:
```diff
@@ -0,0 +1,502 @@
+	case <-state.FinishChan:
+		return false
+	case state.ResultChan <- result:
+		if result.RetrievalResult != nil {
```
Will there ever be a scenario in which result.RetrievalResult _is_ nil?

rvagg commented on pkg/retriever/prioritywaitqueue/prioritywaitqueue.go:
```diff
@@ -0,0 +1,94 @@
+// PriorityWaitQueue is a blocking queue for coordinating goroutines, providing
+// a gating mechanism such that only one goroutine may run at a time, where the
+// goroutine allowed to run is chosen based on a priority comparison function.
+type PriorityWaitQueue[T interface{}] interface {
```
it might look like overkill but this was so handy for writing tests; my tests all use `int` for this so don't need to get tied up with `QueryResponse`

rvagg commented:
Reworked this pretty heavily in the latest commit based on feedback.

* the whole `CidRetrieval` and `retrievalState` thing has been changed, there's just a simple `RetrievalConfig` object which is optional and has all optional properties on it (instrumentation and the various configurable items we use). I think the private/public stuff should be clearer and more idiomatic now?
* There's now a single `Retrieve()` static function that takes your config and a CID to fetch
* There's now a big 'ol `retrieveCandidate()` static function that takes a CID & SP ID and performs the full retrieval flow over them; I bundled it all into one function rather than splitting out the stages just to make it clear that it's a single flow—this could be pulled apart again, it's quite large but that might make sense in a separate file. I really just wanted to demonstrate that flow is a single unit that can happen fully in a goroutine.
* A bit less code overall

kylehuntsman commented on pkg/retriever/retrieve.go:
```diff
@@ -0,0 +1,438 @@
+
+	if queryResponse != nil {
+		// if query is successful, then wait for priority and execute retrieval
+		done = retrieval.waitQueue.Wait(queryResponse)
+		if retrieval.canSendResult() {
+			stats, err = retrievalPhase(ctx, cfg, client, timeout, candidate, queryResponse)
+		}
```
As much as it pains me to have the sender goroutine call close, I like the way this ended up _significantly_ more than the previous iterations because it reads much better. Calling the functions by phases helped as well I think.

kylehuntsman reviewed (approved):
Overall I like this approach a lot more than previous iterations. Naming the functions by phase made it a lot more clear what part of the retrieval I was looking at. The static functions make it clear what is required of that function. The functions turned out relatively flat with very little jumps back and forth between functions. LGTM.

hannahhoward reviewed (approved):
I'm feeling pretty LGTM on this at this point. It's looking really nice.

rvagg commented:
Added a test to demonstrate proper queryresponse priorisation when we get multiple successful queries backed up and we have to choose the "best" one to go next. This is already tested elsewhere in pieces but the new test shows the full flow.

Also rebased to master.
